### ISSUE TITLE
ElementX doesn't support TLS 1.3

### ISSUE THREAD

**Issue by singingtelegram** [vector-im/element-x-ios#786]
### Steps to reproduce

Attempting to login on a TLS v1.3 only server.

#### What did you expect?
The authentication process continues.

#### What happened instead?
log: 🚨 AuthenticationServiceProxy.configure():60 Failed configuring a server: Generic(message: "An error occurred: error sending request for url (https://[hostname]/.well-known/matrix/client): error trying to connect: bad protocol version")

### Your phone model

iPhone 12

### Operating system version

16.4.1

### Application version

1.0.24 (43)

### Homeserver

ocf.berkeley.edu

### Will you send logs?

No

**Comment by jplatte** [comment]
We use the system's native TLS library, could it be that it's just OpenSSL (?) on your system being too old? There's the possibility of bundling [rustls](https://github.com/rustls/rustls) but it will increase the binary size (not sure how much).

**Comment by yu-re-ka** [comment]
The Rust native-tls crate uses the Secure Transport API on iOS.

https://support.apple.com/guide/security/tls-security-sec100a75d12/web

> TLS clients using the SecureTransport APIs can’t use TLS 1.3.

**Comment by manuroe** [comment]
@singingtelegram does legacy EI work on your server?

**Comment by singingtelegram** [comment]
yes, the Element app does work with our server.

**Comment by fuomag9** [comment]
For anyone encountering this, I needed to set tls 1.2 on ALL my servers (domain with the .well-known, actual matrix server and proxy as well)

On caddy that can be done via

```
# Temp fix because of https://github.com/vector-im/element-x-ios/issues/786
tls {
protocols tls1.2 tls1.2
}
```

**Comment by csett86** [comment]
It seems that only Rusts native-tls crate does not support TLS 1.3: https://github.com/sfackler/rust-native-tls/issues/140, while the underlying security-framework crate supports it (https://docs.rs/security-framework/latest/security_framework/secure_transport/struct.SslProtocol.html#associatedconstant.TLS13), because Apples underlying Secure Transport support it since a while now: https://developer.apple.com/documentation/security/tls_protocol_version_t/tls_protocol_version_tlsv13

Also Element iOS works fine with TLS 1.3 only homeservers.

**Comment by pixlwave** [comment]
Looks like there's a second PR to enable this now too… https://github.com/sfackler/rust-native-tls/pull/278

**Comment by csett86** [comment]
Looks like I had a thinko on this and while iOS supported TLS 1.3 for a while eg via URLSession etc, the specific Secure Transport Framework (which is [deprecated](https://developer.apple.com/documentation/security/secure_transport)) does not support TLS 1.3 (https://github.com/sfackler/rust-native-tls/blob/b81f70295bdb6672dd0b5c4686673c0775968d1d/src/imp/security_framework.rs#L53. Sorry for the confusion on my side here.

So the [new PR](https://github.com/sfackler/rust-native-tls/pull/278) for rust-native-tls will not help us here on iOS/macOS, **as it wont bring TLS 1.3 to iOS**.

As Element X Android is already using rustls instead of native-tls (if I read https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/Cargo.toml#L69 correctly), and native-tls will not be able to provide TLS 1.3 for the forseeable future, how do you feel about using rustls on iOS as well? I would be happy to send PRs and be part of testing, I can provide a TLS1.3-only homeserver as well

**Comment by csett86** [comment]
**TLDR: use rustls on iOS/macOS in the matrix-rust-sdk Swift version as Apple recommends to use your own TLS implementation if you use cross-platform code via sockets.**

Long version:

I did more digging into this, and if you look long enough, there is also an [official recommendation](https://developer.apple.com/documentation/technotes/tn3151-choosing-the-right-networking-api#TLS-best-practices) from Apple, that boils down to this if you use cross-platform code with sockets (as we do with the rust sdk):

> To use TLS in that case [BSD Sockets], add your own TLS implementation.

> Don’t use [Secure Transport](https://developer.apple.com/documentation/security/secure_transport) for your TLS implementation. It’s been deprecated since 2019 (see [Versions](https://developer.apple.com/documentation/technotes/tn3151-choosing-the-right-networking-api#Versions)) and doesn’t support TLS 1.3. If you have existing code that uses Secure Transport, make a plan to migrate off it.

Modern TLS implementations including TLS 1.3 on macOS are only available as a built-on via the Apple-specific URLSession / Network framework APIs, so APIs where you feed in a URL and get the response back. They are not available in combination with a generic sockets-based cross-platform code.

With that in mind, there is currently no hope that rust-native-tls would support TLS 1.3 in the forseeable future as there is simply no native TLS implementation in current macOS/iOS that could be used by rust-native-tls.

I'll prepare a PR to switch to rustls for the Swift version of the matrix-rust-sdk.

**Comment by pixlwave** [comment]
@csett86 Sorry for the delay replying. Your message came in right at the point where we discovered that Element X Android wasn't working with custom certificates installed through the OS whereas Element X iOS was (because of Native TLS). So there were discussions about moving EXA to native-tls which was the exact opposite of what you were suggesting. We haven't come to a concrete conclusion on that discussion yet, so for now I think it would be best to leave things as they are.

**Comment by jplatte** [comment]
That's actually easy to fix! Just switch from reqwest's `rustls-tls` feature to `rustls-tls-native-roots`.

**Comment by csett86** [comment]
Unfortunately its not yet that simple, as the underlying rustls-native-certs does not support iOS or Android, if I read this correctly: https://github.com/rustls/rustls-native-certs/issues/3

But the new https://github.com/rustls/rustls-platform-verifier should be able to solve that long-term, both for iOS and Android, once that is supported by reqwest at some point

**Comment by pixlwave** [comment]
Linking to https://github.com/seanmonstar/reqwest/issues/2159 for easy access to the current status of this.

**Comment by alexander-potemkin** [comment]
It might make sense to document that in the section of configuring reverse proxies for Synapse, especially given that Element X is now actively pushed forward for everybody's use and Element 'Classic' is almost deprecated.

It took me a few days and a `tcpdump` at the end, to get to the root cause of that.

**Comment by alexander-potemkin** [comment]
it feels important, bumping this up

**Comment by pixlwave** [comment]
It'd probably be more impactful to bump the issue about supporting rustls-platform-verifier in Reqwest as we're blocked on that.

**Comment by alexander-potemkin** [comment]
Makes sense, done.

**Comment by moabeat-berlin** [comment]
> It might make sense to document that in the section of configuring reverse proxies for Synapse, especially given that Element X is now actively pushed forward for everybody's use and Element 'Classic' is almost deprecated.

I've been looking for the root cause of Element X not working literally for months until I found this issue now. And I can see, that it's still not documented in the [reverse proxy configuration documentation](https://github.com/element-hq/synapse/blob/develop/docs/reverse_proxy.md).

It would have been really helpful, if it would have been added there.

**Comment by csett86** [comment]
FYI: reqwest added support for rustls-platform-verifier in the just released version 0.13 (https://github.com/seanmonstar/reqwest/releases/tag/v0.13.0) and also now recommends rustls instead of native-tls (eg. for TLS 1.3 support and better platform cerificate validation). Once we wire this up in the matrix-rust-sdk, TLS 1.3 support should finally be possible with Element X iOS

**Comment by Neustradamus** [comment]
@csett86: Thanks for your comment!

**Comment by Neustradamus** [comment]
Good news rust-native-tls has now the TLS 1.3 support!

Linked to:
- https://github.com/rust-native-tls/rust-native-tls/pull/353
- https://github.com/rust-native-tls/rust-native-tls/pull/291
- https://github.com/rust-native-tls/rust-native-tls/pull/278
- https://github.com/rust-native-tls/rust-native-tls/pull/235
- https://github.com/rust-native-tls/rust-native-tls/pull/159
- https://github.com/rust-native-tls/rust-native-tls/issues/140
- https://github.com/retep998/winapi-rs/pull/963
- https://github.com/retep998/winapi-rs/pull/859
- https://github.com/olback/barium/issues/14
- https://github.com/element-hq/element-x-ios/issues/786
- https://github.com/matrix-org/matrix-rust-sdk/pull/6053
- https://github.com/deltachat/deltachat-ios/issues/1693

**Comment by ShadowRZ** [comment]
@Neustradamus I don't think it can help here, as the PR you referenced explictly said **Apple platforms currently lack TLS 1.3 support.**

Also please refrain from commenting across many linked issues, I couldn't understand your intent to do this.

**Comment by alexander-potemkin** [comment]
> I couldn't understand ...

I would guess, that it's caused by such a long waiting. TLS 1.3 implementation for a security concerned messenger feels quite of use.
Many are waiting for this to be released / happen as well. At least, I do.

From the engineering side, I understand that iOS (and macOS), native-tls uses the SecureTransport backend via the security-framework crate.
As per Apple's documentation: "TLS clients using the SecureTransport APIs can't use TLS 1.3." TLS 1.3 on Apple platforms is only available through the newer Network.framework and NSURLSession APIs (available since iOS 12.2) - and native-tls does not use it.
And the change required is unlikely to happen soon enough.

So, I guess, we've got to wait.

**Comment by stefanceriu** [comment]
Hello everybody, we are now close to finishing up this whole saga (you can follow the whole thing [here](https://github.com/matrix-org/matrix-rust-sdk/pull/6053#discussion_r2881682432)) but would like to do more testing before we ship it in a nightly.

I've created a new test flight group here https://testflight.apple.com/join/cEcxsGmJ where there's a build based on that RustSDK branch available (version 26.0.3.1 (1)). Once you install it the icon should say "TLS 1.3".

Please test it out if you can and let us know how it goes 🙏

**Comment by csett86** [comment]
Thanks a lot @stefanceriu for the test build, I successfully tested it with a TLS 1.3 only homeserver (sync, media, element call) and the same with a TLS 1.2+ homeserver.

Only issue I could notice: Sometimes the app crashed in the background, I sent feedback via TestFlight when prompted. I wanted to also submit in-app feedback, but there I noticed another issue: The "submit issue" errors that it cannot connect to the feedback server.

**Comment by stefanceriu** [comment]
> Thanks a lot [@stefanceriu](https://github.com/stefanceriu) for the test build, I successfully tested it with a TLS 1.3 only homeserver (sync, media, element call) and the same with a TLS 1.2+ homeserver.

Beautiful, thank you for checking! 🥳 

> Only issue I could notice: Sometimes the app crashed in the background, I sent feedback via TestFlight when prompted. I wanted to also submit in-app feedback, but there I noticed another issue: The "submit issue" errors that it cannot connect to the feedback server.

Unfortunately we have issues with holding SQLite database locks for too long on the rust side and the system deciding to kill the app in the background. It's not great but it shouldn't really affect users. We're working on a fix but it's a tad convoluted.

**Comment by SpiGAndromeda** [comment]
Has somebody managed to get a nginx reverse proxy configured to it works with the non-testflight version?

```
ssl_protocols TLSv1.2;
ssl_prefer_server_ciphers off;
```

Doesnt do the trick for me. The TLS 1.3 version of the app works.

**Comment by marwing** [comment]
> Has somebody managed to get a nginx reverse proxy configured to it works with the non-testflight version?

Plenty of people run nginx as their reverse proxy, it seems to be one of the more popular solutions. Not working could be a couple of things, have you checked the logs yet (nginx and app)?

Also, if you want to discuss further, you could ask in https://matrix.to/#/#element-x-ios:matrix.org, so this doesn't spam everyone who subscribed to this issue.

**Comment by pixlwave** [comment]
This was fixed by https://github.com/matrix-org/matrix-rust-sdk/pull/6053 and is included in 26.04.0.

**Comment by Neustradamus** [comment]
@pixlwave: Good news, thanks!
